# 05_숫자 계산하기

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#정수-계산하기" data-toc-modified-id="정수-계산하기-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>정수 계산하기</a></span></li><li><span><a href="#실수-계산하기" data-toc-modified-id="실수-계산하기-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>실수 계산하기</a></span></li></ul></div>

## 정수 계산하기

> 몫과 나머지를 함께 구하기

In [2]:
divmod(5,2) # 앞이 몫, 뒤가 나머지

(2, 1)

In [3]:
# 몫과 나머지를 두 개의 변수에 저장하기
quotient, remainder = divmod(5, 2)
print(quotient, remainder)

2 1


> 진수, 8진수, 16진수

정수는 10진수 이외에도 2진수, 8진수, 16진수로도 표현할 수 있다.

* 2진수 : 숫자 앞에 0b 를 붙이며 0과 1을 사용한다.
* 8진수 : 숫자 앞에 0o (숫자 0과 소문자 o) 를 붙이며 0부터 7까지 사용한다.
* 16진수 : 숫자 앞에 0x 또는 0X 를 붙이며 0부터 9, A부터 F까지 사용한다. (소문자 a부터 f도 가능.)

In [4]:
# 2진수
0b110

6

In [5]:
# 8진수
0o10

8

In [6]:
# 16진수
0xF

15

## 실수 계산하기

In [8]:
4.3 - 2.7

1.5999999999999996

In [9]:
1.5 * 3.1

4.65

In [10]:
5.5 / 3.1

1.7741935483870968

그런데 `4.3 - 2.7`의 결과가 좀 이상하다. 1.6이 나와야 하는데 1.5999999999999996이 나왔다. 왜냐하면 컴퓨터는 실수를 표현할 때 오차가 발생하기 때문!

나중에 실무에서 실수를 다룰 때 오차에 대한 적절한 처리가 필요하다.

> 실수 값의 오차

파이썬에서 0.1 + 0.2의 값은 0.3이 나올 것 같지만 실제로는 0.30000000000000004이 나온다.

파이썬은 실수를 부동소수점 방식으로 표현하는데 부동소수점은 실수를 정확히 표현할 수 없는 문제가 있다. 그래서 0.3이 아닌 0.30000000000000004가 나온다.

In [11]:
0.1 + 0.2

0.30000000000000004

> 두 실수가 같은지 판단하기

만약 두 실수가 같은지 판단할때는 ==을 사용하면 안된다.

In [12]:
0.1 + 0.2 == 0.3

False

0.1 + 0.2는 0.30000000000000004 이므로 0.3과는 다르다. 

이렇게 실수를 근삿값으로 표현하면서 발생하는 문제를 부동소수점 반올림 오차 (rounding error) 라고 한다.

따라서 실수를 비교할 때는 연산한 값과 비교할 값의 차이를 구한뒤 sys.float_info.epsilon보다 작거나 같은지 판단해야 한다.

In [16]:
import math, sys

x = 0.1 + 0.2
math.fabs(x-0.3) <= sys.float_info.epsilon

True

In [14]:
sys.float_info.epsilon

2.220446049250313e-16

x의 값 0.30000000000000004 에서 0.3을 뺀 값이 sys.float_info.epsilon보다 작거나 같으므로 두 값은 같은 값이라 할 수 있다. 
특히 math.fabs를 사용해서 두 값의 차이를 절댓값으로 만들면 음수가 나오더라도 정상적으로 판단할 수 있다.

여기서 sys.float_info.epsilon에 저장된 값을 머신 엡실론 (machine epsilon)이라고 부르는데, 어떤 실수를 가장 가까운 부동소수점 실수로 반올림했을 때 상대 오차는 항상 머신 엡실론 이하이다. 즉, 머신 엡실론은 반올림 오차의 상한값이며 연산한 값과 비교할 값의 차이가 머신 엡실론보다 작거나 같다면 두 실수는 같은 값이라고 할 수 있다.

파이썬 3.5이상부터는 두 실수가 같은지 판단할 때 math.insclose 함수를 사용하면 된다.

In [17]:
import math

math.isclose(0.1+0.2, 0.3)

True

> Decimal으로 정확한 자릿수 표현하기

반올림 오차가 없는 고정소수점을 사용하려면 decimal 모듈의 Decimal을 사용하면 된다. Decimal은 숫자를 10진수로 처리하여 정확한 소수점 자릿수를 표현한다.

In [18]:
from decimal import Decimal

Decimal('0.1') + Decimal('0.2')

Decimal('0.3')

> Fraction으로 분수 표현하기

순환소수는 고정소수점이라도 정확히 표현할 수 없다. 이때는 fractions 모듈의 Fraction을 사용하여 분수로 표현하면 된다.

In [19]:
from fractions import Fraction

Fraction('10/3') # 10을 3으로 나누면 순환소수 3.33333... 이지만 분수 3분의 10으로 표현

Fraction(10, 3)

> 복소수

파이썬에서는 실수부와 허수부로 이루어진 복소수 (complex number) 도 사용할 수 있다. 이 때 허수부는 숫자 뒤에 j를 붙인다. (수학에서는 허수를 i로 표현하지만 공학에서는 j를 사용한다.)

In [20]:
1.2 + 1.3j

(1.2+1.3j)

두 실수를 복소수로 만들 때는 complex를 사용하면 된다.

In [21]:
complex(1.2, 1.3)

(1.2+1.3j)